rhizome/host/lib/vm_setup.rb

```diff
@@ -335,6 +338,29 @@ def routes4(ip4, ip4_local, nics)
     end
   end
 
+  def update_via_routes(nics)
+    # we create tap devices in "interfaces" function in this file. but
+    # code execution happens faster than linux taking care of the device creation.
+    # that's why by the time we reach this function, we need to check whether the
+    # device is created or not and then proceed to modify the routes.
+    success = false
+    5.times do
+      if r("ip -n #{q_vm} link | grep -E '^[0-9]+: nc[^:]+:' | grep -q 'state UP' && echo UP || echo DOWN").chomp == "UP"
+        success = true
+        break
+      end
+      sleep 0.1
+    end
+    unless success
+      raise "VM #{q_vm} tap device not ready after 5 retries."
+    end
+
+    nics.each do |nic|
+      local_ip4 = NetAddr::IPv4Net.parse(nic.net4)
+      r "ip -n #{q_vm} route replace #{local_ip4.to_s.shellescape} via #{local_ip4.nth(0).to_s.shellescape} dev #{nic.tap}"
```

Comment:
sometimes, using the 0th index of a network as a hop (or indeed, for anything) has caused some problems, especially with older linux... @furkansahin has the story. Maybe it should be nth(1)?  cc @furkansahin

Reply:
Mostly, the issue is it has not been traditional to use the all-zeros suffix to do anything but identify the network, even though theoretically, there's no reason it can't be used. Old linux abided tradition, new linux relaxed it and made it more general, abiding theory.